extensions/config-yaml/runtime/src/main/java/io/quarkus/yaml/configuration/runtime/YamlConfigSource.java

```diff
@@ -0,0 +1,127 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+import org.yaml.snakeyaml.Yaml;
+
+public abstract class YamlConfigSource implements ConfigSource, Serializable {
+
+    private static final long serialVersionUID = -875865410658150582L;
+
+    private static final String CONFIG_ORDINAL_KEY = "config_ordinal";
+    private static final String CONFIG_ORDINAL_DEFAULT_VALUE = "100";
+    private static final String PROPERTY_KEY_SEPARATOR = ".";
+    private static final String COMMA = ",";
+    private static final String ESCAPED_COMMA = "\\,";
+
+    private final Map<String, String> properties;
+    private final int ordinal;
+
+    protected final String source;
+
+    protected YamlConfigSource(String source, int ordinal) {
+        this.source = source;
+        this.properties = buildConfigPropertiesFromYaml();
+        if (properties.containsKey(CONFIG_ORDINAL_KEY)) {
+            this.ordinal = Integer.parseInt(properties.getOrDefault(CONFIG_ORDINAL_KEY, CONFIG_ORDINAL_DEFAULT_VALUE));
+        } else {
+            this.ordinal = ordinal;
+        }
+    }
+
+    protected abstract InputStream openYamlFileStream();
+
+    private Map<String, String> buildConfigPropertiesFromYaml() {
+        try (InputStream is = openYamlFileStream()) {
+            if (is == null) {
+                return Collections.emptyMap();
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> yamlContent = new Yaml().loadAs(is, HashMap.class);
+            Map<String, String> configProperties = new HashMap<>();
+            for (YamlConfigProperty prop : convertYamlMapToProperties("", yamlContent)) { // The root `baseKey` is empty.
+                configProperties.put(prop.key, prop.value);
+            }
+            return configProperties;
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<YamlConfigProperty> convertYamlMapToProperties(String baseKey, Object object) {
+        if (object instanceof Map) {
+            List<YamlConfigProperty> properties = new ArrayList<>();
+            for (Entry<?, ?> entry : ((Map<?, ?>) object).entrySet()) {
+                String entryKey = baseKey + (baseKey.isEmpty() ? "" : PROPERTY_KEY_SEPARATOR) + entry.getKey();
+                properties.addAll(convertYamlMapToProperties(entryKey, entry.getValue()));
+            }
+            return properties;
+        } else {
+            String value;
+            if (object instanceof List) {
+                value = stringifyList(object);
+            } else {
+                value = String.valueOf(object);
+            }
+            return Collections.singletonList(new YamlConfigProperty(baseKey, value));
+        }
+    }
+
+    private String stringifyList(Object list) {
```

Comment:
@gwenneg just a minor comment - could we build the string inside the loop and avoid `line 89` (looping twice)?

Reply:
Sure, I've replaced the `List` with a `StringJoiner`, thanks!